AxonDeepSeg/morphometrics/compute_morphometrics.py

```diff
@@ -1,197 +1,251 @@
+        # distance metric will be more useful for the watershed algorithm below.
+        distance = ndi.distance_transform_edt(im_axon)
+        # local_maxi = feature.peak_local_max(distance, indices=False, footprint=np.ones((31, 31)), labels=axonmyelin)
+
+        # Get axon centroid as int (not float) to be used as index
+        ind_centroid = ([int(props.centroid[0]) for props in axon_objects],
+                        [int(props.centroid[1]) for props in axon_objects])
+
+        # Create an image with axon centroids, which value corresponds to the value of the axon object
+        im_centroid = np.zeros_like(im_axon)
+        for i in range(len(ind_centroid[0])):
+            # Note: The value "i" corresponds to the label number of im_axon_label
+            im_centroid[ind_centroid[0][i], ind_centroid[1][i]] = i
+
+        # markers = ndi.label(local_maxi)[0]
+        # Watershed segmentation of axonmyelin using distance map
+        im_axonmyelin_label = morphology.watershed(-distance, im_centroid, mask=im_axonmyelin)
+        # Measure properties of each axonmyelin object
+        axonmyelin_objects = measure.regionprops(im_axonmyelin_label)
+
+    # DEBUG
+    # from matplotlib import colors
+    # from matplotlib.pylab import *
+    # random_cmap = matplotlib.colors.ListedColormap(np.random.rand(256, 3))
+    # matshow(im_axon_label, fignum=1, cmap=random_cmap), show()
+    # import datetime
+    # savefig('fig_' + datetime.datetime.strftime(datetime.datetime.now(), '%Y%m%d%H%M%S%f') + '.png', format='png',
+    #         transparent=True, dpi=100)
+
+    # Loop across axon property and fill up dictionary with morphometrics of interest
+    for prop_axon in axon_objects:
+        # Centroid
+        y0, x0 = prop_axon.centroid
+        # Solidity
+        solidity = prop_axon.solidity
+        # Eccentricity
+        eccentricity = prop_axon.eccentricity
+        # Axon equivalent diameter in micrometers
+        axon_diam = prop_axon.equivalent_diameter * pixelsize
+        # Axon area in µm^2
+        axon_area = prop_axon.area * (pixelsize ** 2)
+        # Axon orientation angle
+        orientation = prop_axon.orientation
+        # Add metrics to list of dictionaries
+        stats = {'y0': y0,
+                 'x0': x0,
+                 'axon_diam': axon_diam,
+                 'axon_area': axon_area,
+                 'solidity': solidity,
+                 'eccentricity': eccentricity,
+                 'orientation': orientation}
+        # Deal with myelin
+        if im_myelin is not None:
+            # Find label of axonmyelin corresponding to axon centroid
+            label_axonmyelin = im_axonmyelin_label[int(y0), int(x0)]
+            # TODO: use logger
+            # print(label_axonmyelin)
+            # print('x, y = {}, {}'.format(x0, y0))
+            if label_axonmyelin:
+                # Get corresponding index from axonmyelin_objects list
+                ind_axonmyelin = \
```

Comment:
@jcohenadad This index calculation doesn't seem to be implemented well. I tried it for an image with three axon, and by debugging with a `print(ind_axonmyelin)` call right after, it only ever returns `0`.

I found this bug by looking at what the myelin thickness/area/gratio was for each of my axons, and found that myelin thickness area was exactly the same for each one (whereas the gratio was different, but that's because it's also using the axon diameter, which is well implemented & different).

Steps I followed:

```python
import numpy as np
import math
import os
import imageio
from sys import platform as _platform
if _platform == "darwin": # Mac OSX
     import matplotlib as mpl
     mpl.use('TkAgg')
import matplotlib.pyplot as plt

from AxonDeepSeg.morphometrics.compute_morphometrics import *

%matplotlib inline

# Change the image and segmentation paths here for your sample:
path_img = '/Users/mathieuboudreau/Downloads/sample282/image.png'
path_pred = '/Users/mathieuboudreau/Downloads/sample282/AxonDeepSeg.png'

# Read paths and compute axon/myelin masks
img = imageio.imread(path_img)
pred = imageio.imread(path_pred)
pred_axon = pred > 200
pred_myelin = np.logical_and(pred >= 50, pred <= 200)
path_folder, file_name = os.path.split(path_img)

print(stats_array[0]['myelin_diam'])
print(stats_array[1]['myelin_diam'])
print(stats_array[2]['myelin_diam'])

print(stats_array[0]['myelin_area'])
print(stats_array[1]['myelin_area'])
print(stats_array[2]['myelin_area'])

print(stats_array[0]['gratio'])
print(stats_array[1]['gratio'])
print(stats_array[2]['gratio'])

# Get axon diameter value of axon object 0
print(stats_array[0]['axon_diam'])
print(stats_array[1]['axon_diam'])
print(stats_array[2]['axon_diam'])
```

Output: 
<img width="916" alt="Capture d’écran 2019-03-09 à 17 51 00" src="https://user-images.githubusercontent.com/1421029/54077509-12d43180-4294-11e9-90ef-c95eb55a9993.png">

Segmented mask:
![AxonDeepSeg](https://user-images.githubusercontent.com/1421029/54077505-0e0f7d80-4294-11e9-8d89-b6cd49a01030.png)

Reply:
This might also explain the high variance in your listed above g-values.

Reply:
@jcohenadad some additional details found through debugging

1) I think the following line is supposed to return the index of the myelin object (e.g. 0,1, or 2), but it always returns 0:

https://github.com/neuropoly/axondeepseg/blob/b3577b26945d2edbddd31e2045bf9b9d953e5872/AxonDeepSeg/morphometrics/compute_morphometrics.py#L117-L119

2) `label_axonmyelin` is always equal to 1 (found through debugging/`print()`).

3) Reminder that `label_axonmyelin` is defined as: 

https://github.com/neuropoly/axondeepseg/blob/b3577b26945d2edbddd31e2045bf9b9d953e5872/AxonDeepSeg/morphometrics/compute_morphometrics.py#L112

4) It appears that what the above was meant to return (I think) is the label of the axonmyelin object (e.g. 0,1,2; or something equivelent ?), however, it always returns `1`, because `im_axonmyelin_label` is a binary mask of the image (0 or 1). This is defined here: 

https://github.com/neuropoly/axondeepseg/blob/b3577b26945d2edbddd31e2045bf9b9d953e5872/AxonDeepSeg/morphometrics/compute_morphometrics.py#L73-L74

Reply:
It appears that the second parameter of `watershed`, the `markers` argument (here defined as `im_centroid`), should be an array of the same size as the `image` (here `-distance`), with the different integer values (the "markers). However here, `im_centroid` is a tuple of arrays of the centroid coordinates (`([724, 1260, 1517], [1115, 400, 1054])`). Maybe that's the issue?

Source for `watershed` argument definitions: http://scikit-image.org/docs/0.7.0/api/skimage.morphology.watershed.html#skimage.morphology.watershed.watershed

Reply:
Omit my last (deleted) message; I was printing the `ind_centroid` and not `im_centroid`.

Reply:
Getting closer. I think (part of) the bug lies here:

https://github.com/neuropoly/axondeepseg/blob/b3577b26945d2edbddd31e2045bf9b9d953e5872/AxonDeepSeg/morphometrics/compute_morphometrics.py#L66-L67

`np.zeros_like` copies both the shape **and** type, so since `im_axon` is a bool, so was `im_centroid`, even though later we wanted to assign it labels that are integers greater than 1 (e.g. instead of (0,1,2,3,...), the labels were being overriden by the type to (0,1,1,1...).

Reply:
Doing ` im_centroid = np.zeros_like(im_axon, dtype=uint8)` solves the first problem I flag. For the simulated image below, I now get all different myelin areas and thicknesses for the 3 out of 4 axons that were identified. For the actual axon image above, since the three axons are touching, I think only two were correctly separated by the watershed, since two of the diameters are identical but one is different (improvement!).

<img width="1138" alt="Capture d’écran 2019-03-09 à 23 46 37" src="https://user-images.githubusercontent.com/1421029/54080217-a4a86280-42c5-11e9-8ca0-ccf35435e0dd.png">

Reply:
Yep, only two `axonmyelin_objects` were identified using that watershed run:

![im_axonmyelin_label](https://user-images.githubusercontent.com/1421029/54080300-4d0af680-42c7-11e9-8f9f-e80f3c766e18.png)

Reply:
I know there's likely not an easy way to ensuring that these are always separated perfectly, but maybe we should make a condition where if `axonmyelin` object centroids are too far from the `axon` object centroids, to omit these automatically (because otherwise, the metrics will get screwed up for a variety of reasons), maybe defining them (the myelin metrics) as `Null` for these cases. What do you think @jcohenadad ?

Reply:
Actually; by simply resolving the issue in my comment below (i.e. setting `im_centroid[ind_centroid[0][i], ind_centroid[1][i]] = i + 1 `), the axons (strangely) now appear to have different values:

![im_axonmyelin_label](https://user-images.githubusercontent.com/1421029/54080486-6c0b8780-42cb-11e9-93a7-661363f8935b.png)

<img width="1138" alt="Capture d’écran 2019-03-10 à 00 29 11" src="https://user-images.githubusercontent.com/1421029/54080488-8e050a00-42cb-11e9-8a7b-0716e3cf9240.png">